File: check_models.js
#!/usr/bin/env node
/**
 * Model availability checker for GitHub Actions.
 * Tests Deep.Assistant API models and reports their availability.
 */
import OpenAI from 'openai';
import * as fs from 'fs';

// Model configurations
const MODELS = [
  "o3-mini",
  "o1-preview",
  "o1-mini",
  "gpt-4o",
  "gpt-4o-mini",
  "gpt-3.5-turbo",
  "gpt-auto",
  "claude-3-opus",
  "claude-3-5-sonnet",
  "claude-3-5-haiku",
  "claude-3-7-sonnet",
  "deepseek-chat",
  "deepseek-reasoner",
];

/**
 * Test a single model for availability
 */
async function testModel(client, model) {
  const result = {
    model,
    available: false,
    status: '',
    error: null
  };

  try {
    const completion = await client.chat.completions.create({
      messages: [{ role: 'user', content: 'Test' }],
      model,
      max_tokens: 10,
    });

    const responseModel = completion.model;

    // Check if the returned model matches the requested model
    if (responseModel === model) {
      result.available = true;
      result.status = 'Available';
    } else {
      result.status = `Wrong model returned: ${responseModel}`;
      result.error = `Expected ${model}, got ${responseModel}`;
    }
  } catch (error) {
    result.status = 'Error';

    if (error.status) {
      result.status = `HTTP ${error.status}`;
      result.error = error.message || `HTTP error ${error.status}`;
    } else if (error.code === 'ETIMEDOUT' || error.message?.includes('timeout')) {
      result.status = 'Timeout';
      result.error = 'Request timed out';
    } else {
      result.error = error.message || String(error);
    }
  }

  return result;
}

/**
 * Test all models or a specific subset
 */
async function testAllModels(apiKey, modelsToTest = null) {
  const models = modelsToTest || MODELS;

  console.log(`Testing ${models.length} model(s)...`);

  const client = new OpenAI({
    apiKey,
    baseURL: process.env.OPENAI_API_BASE || 'https://api.deep.assistant.run.place/v1',
    timeout: 60000, // 60 second timeout
  });

  // Test all models in parallel
  const results = await Promise.all(
    models.map(model => testModel(client, model))
  );

  // Organize results
  const available = [];
  const unavailable = [];

  for (const result of results) {
    if (result.available) {
      available.push(result.model);
    } else {
      unavailable.push({
        model: result.model,
        status: result.status,
        error: result.error
      });
    }
  }

  return {
    total: models.length,
    available,
    unavailable,
    results
  };
}

/**
 * Print test results in a readable format
 */
function printResults(summary) {
  console.log('\n' + '='.repeat(60));
  console.log('MODEL AVAILABILITY TEST RESULTS');
  console.log('='.repeat(60));

  console.log(`\nTotal models tested: ${summary.total}`);
  console.log(`Available models: ${summary.available.length}`);
  console.log(`Unavailable models: ${summary.unavailable.length}`);

  if (summary.available.length > 0) {
    console.log('\n✓ AVAILABLE MODELS:');
    for (const model of summary.available) {
      console.log(`  - ${model}`);
    }
  }

  if (summary.unavailable.length > 0) {
    console.log('\n✗ UNAVAILABLE MODELS:');
    for (const item of summary.unavailable) {
      console.log(`  - ${item.model}: ${item.status}`);
      if (item.error) {
        console.log(`    Error: ${item.error}`);
      }
    }
  }

  console.log('\n' + '='.repeat(60));
}

/**
 * Main entry point
 */
async function main() {
  // Get API key from environment
  const apiKey = process.env.OPENAI_API_KEY;
  if (!apiKey) {
    console.error('Error: OPENAI_API_KEY environment variable not set');
    process.exit(1);
  }

  // Parse command line arguments for specific models
  const modelsToTest = process.argv.length > 2 ? process.argv.slice(2) : MODELS;

  if (process.argv.length > 2) {
    console.log(`Testing specific models: ${modelsToTest.join(', ')}`);
  }

  // Run tests
  const summary = await testAllModels(apiKey, modelsToTest);

  // Print results
  printResults(summary);

  // Output for GitHub Actions
  if (process.env.GITHUB_ACTIONS) {
    const outputFile = process.env.GITHUB_OUTPUT;
    if (outputFile) {
      const output = [
        `available_count=${summary.available.length}`,
        `unavailable_count=${summary.unavailable.length}`,
        `total_count=${summary.total}`
      ].join('\n') + '\n';

      fs.appendFileSync(outputFile, output);
    }
  }

  // Exit with error code if any models are unavailable
  if (summary.unavailable.length > 0) {
    console.log(`\n⚠ Warning: ${summary.unavailable.length} model(s) unavailable`);
    process.exit(1);
  } else {
    console.log('\n✓ All models are available!');
    process.exit(0);
  }
}

main().catch(error => {
  console.error('Fatal error:', error);
  process.exit(1);
});

File: check_models.py
#!/usr/bin/env python3
"""
Model availability checker for GitHub Actions.
Tests Deep.Assistant API models and reports their availability.
"""
import os
import sys
import json
import asyncio
import aiohttp
from typing import Dict, List, Any


# Model configurations
MODELS = [
    "o3-mini",
    "o1-preview",
    "o1-mini",
    "gpt-4o",
    "gpt-4o-mini",
    "gpt-3.5-turbo",
    "gpt-auto",
    "claude-3-opus",
    "claude-3-5-sonnet",
    "claude-3-5-haiku",
    "claude-3-7-sonnet",
    "deepseek-chat",
    "deepseek-reasoner",
]


async def test_model(session: aiohttp.ClientSession, model: str, api_key: str) -> Dict[str, Any]:
    """Test a single model for availability."""
    url = f"{os.getenv('OPENAI_API_BASE', 'https://api.deep.assistant.run.place/v1')}/chat/completions"

    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json"
    }

    payload = {
        "messages": [{"role": "user", "content": "Test"}],
        "model": model,
        "max_tokens": 10
    }

    result = {
        "model": model,
        "available": False,
        "status": "",
        "error": None
    }

    try:
        async with session.post(
            url,
            json=payload,
            headers=headers,
            timeout=aiohttp.ClientTimeout(total=60)
        ) as response:
            if response.status == 200:
                data = await response.json()
                response_model = data.get("model", "")

                # Check if the returned model matches the requested model
                if response_model == model:
                    result["available"] = True
                    result["status"] = "Available"
                else:
                    result["status"] = f"Wrong model returned: {response_model}"
                    result["error"] = f"Expected {model}, got {response_model}"
            else:
                result["status"] = f"HTTP {response.status}"
                try:
                    error_data = await response.json()
                    if 'error' in error_data and 'message' in error_data['error']:
                        result["error"] = error_data['error']['message']
                        result["status"] = f"HTTP {response.status}: {result['error']}"
                except:
                    result["error"] = f"HTTP error {response.status}"

    except asyncio.TimeoutError:
        result["status"] = "Timeout"
        result["error"] = "Request timed out after 60 seconds"
    except Exception as e:
        result["status"] = "Error"
        result["error"] = str(e)

    return result


async def test_all_models(api_key: str, models: List[str] = None) -> Dict[str, Any]:
    """Test all models or a specific subset."""
    if models is None:
        models = MODELS

    print(f"Testing {len(models)} model(s)...")

    async with aiohttp.ClientSession() as session:
        tasks = [test_model(session, model, api_key) for model in models]
        results = await asyncio.gather(*tasks)

    # Organize results
    available = []
    unavailable = []

    for result in results:
        if result["available"]:
            available.append(result["model"])
        else:
            unavailable.append({
                "model": result["model"],
                "status": result["status"],
                "error": result["error"]
            })

    return {
        "total": len(models),
        "available": available,
        "unavailable": unavailable,
        "results": results
    }


def print_results(summary: Dict[str, Any]) -> None:
    """Print test results in a readable format."""
    print("\n" + "=" * 60)
    print("MODEL AVAILABILITY TEST RESULTS")
    print("=" * 60)

    print(f"\nTotal models tested: {summary['total']}")
    print(f"Available models: {len(summary['available'])}")
    print(f"Unavailable models: {len(summary['unavailable'])}")

    if summary['available']:
        print("\n✓ AVAILABLE MODELS:")
        for model in summary['available']:
            print(f"  - {model}")

    if summary['unavailable']:
        print("\n✗ UNAVAILABLE MODELS:")
        for item in summary['unavailable']:
            print(f"  - {item['model']}: {item['status']}")
            if item['error']:
                print(f"    Error: {item['error']}")

    print("\n" + "=" * 60)


def main():
    """Main entry point."""
    # Get API key from environment
    api_key = os.getenv("OPENAI_API_KEY")
    if not api_key:
        print("Error: OPENAI_API_KEY environment variable not set", file=sys.stderr)
        sys.exit(1)

    # Parse command line arguments for specific models
    models_to_test = MODELS
    if len(sys.argv) > 1:
        # Test specific models passed as arguments
        models_to_test = sys.argv[1:]
        print(f"Testing specific models: {', '.join(models_to_test)}")

    # Run tests
    summary = asyncio.run(test_all_models(api_key, models_to_test))

    # Print results
    print_results(summary)

    # Output JSON for GitHub Actions
    if os.getenv("GITHUB_ACTIONS"):
        output_file = os.getenv("GITHUB_OUTPUT")
        if output_file:
            with open(output_file, "a") as f:
                f.write(f"available_count={len(summary['available'])}\n")
                f.write(f"unavailable_count={len(summary['unavailable'])}\n")
                f.write(f"total_count={summary['total']}\n")

    # Exit with error code if any models are unavailable
    if summary['unavailable']:
        print(f"\n⚠ Warning: {len(summary['unavailable'])} model(s) unavailable")
        sys.exit(1)
    else:
        print("\n✓ All models are available!")
        sys.exit(0)


if __name__ == "__main__":
    main()
